crates/viewer/re_space_view_dataframe/src/expanded_rows.rs

```diff
@@ -0,0 +1,136 @@
+    valid_for: egui::Id,
+}
+
+impl Default for ExpandedRowsCache {
+    fn default() -> Self {
+        Self {
+            expanded_rows: BTreeMap::default(),
+            valid_for: egui::Id::new(""),
+        }
+    }
+}
+
+impl ExpandedRowsCache {
+    /// This sets the query used for cache invalidation.
+    ///
+    /// If the query doesn't match the cached one, the state will be reset.
+    fn validate_id(&mut self, id: egui::Id) {
+        if id != self.valid_for {
+            self.valid_for = id;
+            self.expanded_rows = BTreeMap::default();
+        }
+    }
+}
+
+/// Helper to keep track of row expansion.
+///
+/// This is a short-lived struct to be created every frame. The persistent state is stored in
+/// [`ExpandedRowsCache`].
+///
+/// Uses egui's animation support to animate the row expansion/contraction. For this to work:
+/// - When collapsed, the row entry must be set to 0 instead of being removed. Otherwise, it will no
+///   longer be "seen" by the animation code. Technically, it could be removed _after_ the
+///   animation completes, but it's not worth the complexity.
+/// - When the row is first expanded, for the animation to work, it must be immediately seeded to 0
+///   for the animation to have a starting point.
+pub(crate) struct ExpandedRows<'a> {
+    /// Base row height.
+    row_height: f32,
+
+    /// Cache containing the row expanded-ness.
+    cache: &'a mut ExpandedRowsCache,
+
+    /// [`egui::Context`] used to animate the row expansion.
+    egui_ctx: egui::Context,
+
+    /// [`egui::Id`] used to store the animation state.
+    id: egui::Id,
+}
+
+impl<'a> ExpandedRows<'a> {
+    /// Create a new [`ExpandedRows`] instance.
+    ///
+    /// `egui_ctx` is used to animate the row expansion
+    /// `id` is used to store the animation state and invalidate the cache, make it persistent and
+    /// unique
+    pub(crate) fn new(
+        egui_ctx: egui::Context,
+        id: egui::Id,
+        cache: &'a mut ExpandedRowsCache,
+        row_height: f32,
+    ) -> Self {
+        // (in-)validate the cache
+        cache.validate_id(id);
+
+        Self {
+            row_height,
+            cache,
+            egui_ctx,
+            id,
+        }
+    }
+
+    /// Implementation for [`egui_table::TableDelegate::row_top_offset`].
+    pub(crate) fn row_top_offset(&self, row_nr: u64) -> f32 {
+        self.cache
+            .expanded_rows
+            .range(0..row_nr)
+            .map(|(expanded_row_nr, expanded)| {
+                self.egui_ctx.animate_value_with_time(
+                    self.row_id(*expanded_row_nr),
+                    *expanded as f32 * self.row_height,
+                    self.egui_ctx.style().animation_time,
+                )
+            })
+            .sum::<f32>()
+            + row_nr as f32 * self.row_height
+    }
+
+    /// Return by how much row space this row is expended.
+    pub(crate) fn row_expansion(&self, row_nr: u64) -> u64 {
```

Comment:
I don't understand this docstring.

Does this return how many rows the row expands too?

I think we should introduce a second ford for these "sub-rows", like "like". As in: "Most table rows consist of a single line, but rows can sometimes be expanded to cover multiple lines"